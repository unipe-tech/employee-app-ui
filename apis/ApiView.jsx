import React from "react";
import { Button } from "@react-native-material/core";
import TextButton from "../components/atoms/TextButton";

const ApiView = (props) => {
  const { disabled, goForFetch, loading, style, title } = props;

  return (
    <TextButton
      label={loading ? "Verifying" : "Continue"}
      uppercase={false}
      type="solid"
      color="#4E46F1"
      style={style}
      disabled={loading || disabled}
      onPress={() => {
        goForFetch();
      }}
    />
  );
};

export default ApiView;
