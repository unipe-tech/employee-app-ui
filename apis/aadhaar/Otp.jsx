import { useEffect, useState } from "react";
import { Alert } from "react-native";
import { useDispatch, useSelector } from "react-redux";
import { OG_API_KEY } from "@env";
import { useNavigation } from "@react-navigation/core";
import Analytics from "appcenter-analytics";

import PrimaryButton from "../../components/PrimaryButton";
import { aadhaarBackendPush } from "../../helpers/BackendPush";
import { KYC_AADHAAR_GENERATE_OTP_API_URL } from "../../services/constants";
import {
  addSubmitOTPtxnId,
  addVerifyMsg,
  addVerifyStatus,
  addVerifyTimestamp,
} from "../../store/slices/aadhaarSlice";
import { resetTimer } from "../../store/slices/timerSlice";

const AadhaarOtpApi = (props) => {
  const dispatch = useDispatch();
  const navigation = useNavigation();

  const [loading, setLoading] = useState(false);
  const [backendPush, setBackendPush] = useState(false);

  const id = useSelector((state) => state.auth.id);
  const aadhaarSlice = useSelector((state) => state.aadhaar);
  const [submitOTPtxnId, setSubmitOTPtxnId] = useState(
    aadhaarSlice?.submitOTPtxnId
  );
  const [verifyMsg, setVerifyMsg] = useState(aadhaarSlice?.verifyMsg);
  const [verifyStatus, setVerifyStatus] = useState(aadhaarSlice?.verifyStatus);
  const [verifyTimestamp, setVerifyTimestamp] = useState(
    aadhaarSlice?.verifyTimestamp
  );

  useEffect(() => {
    dispatch(addSubmitOTPtxnId(submitOTPtxnId));
  }, [submitOTPtxnId]);

  useEffect(() => {
    dispatch(addVerifyMsg(verifyMsg));
  }, [verifyMsg]);

  useEffect(() => {
    dispatch(addVerifyStatus(verifyStatus));
  }, [verifyStatus]);

  useEffect(() => {
    dispatch(addVerifyTimestamp(verifyTimestamp));
  }, [verifyTimestamp]);

  useEffect(() => {
    console.log(
      "KYC_AADHAAR_GENERATE_OTP_API_URL: ",
      KYC_AADHAAR_GENERATE_OTP_API_URL
    );
    console.log("AadhaarOtpApi aadhaarSlice: ", aadhaarSlice);
    if (backendPush) {
      aadhaarBackendPush({
        id,
        data: aadhaarSlice?.data,
        number: aadhaarSlice?.number,
        verifyMsg,
        verifyStatus,
        verifyTimestamp,
      });
      setBackendPush(false);
      setLoading(false);
    }
  }, [backendPush]);

  const goForFetch = () => {
    setLoading(true);

    const options = {
      method: "POST",
      headers: {
        "X-Auth-Type": "API-Key",
        "X-API-Key": OG_API_KEY,
        "Content-Type": "application/json",
      },
      body: JSON.stringify(props.data),
    };

    fetch(KYC_AADHAAR_GENERATE_OTP_API_URL, options)
      .then((response) => response.json())
      .then((responseJson) => {
        try {
          if (responseJson.status == "200") {
            switch (responseJson.data.code) {
              case "1001":
                setSubmitOTPtxnId(responseJson.data.transaction_id);
                setVerifyMsg("OTP sent to User");
                setVerifyStatus("PENDING");
                setBackendPush(true);
                setVerifyTimestamp(responseJson.timestamp);
                dispatch(resetTimer());
                Analytics.trackEvent("Aadhaar|Otp|Success", {
                  userId: id,
                });
                {
                  props.type == "KYC"
                    ? navigation.navigate("KYC", {
                        screen: "AADHAAR",
                        params: {
                          screen: "Verify",
                        },
                      })
                    : navigation.navigate("AadhaarVerify");
                }
                break;
              default:
                setVerifyMsg(responseJson.data.message);
                Analytics.trackEvent("Aadhaar|Otp|Error", {
                  userId: id,
                  error: responseJson.data.message,
                });
                setVerifyStatus("ERROR");
                setBackendPush(true);
                Alert.alert("Error", responseJson.data.message);
                break;
            }
          } else if (responseJson?.error?.message) {
            setVerifyMsg(responseJson.error.message);
            Analytics.trackEvent("Aadhaar|Otp|Error", {
              userId: id,
              error: responseJson.error.message,
            });
            setVerifyStatus("ERROR");
            setBackendPush(true);
            Alert.alert("Error", responseJson.error.message);
          } else {
            setVerifyMsg(responseJson.message);
            Analytics.trackEvent("Aadhaar|Otp|Error", {
              userId: id,
              error: responseJson.message,
            });
            setVerifyStatus("ERROR");
            setBackendPush(true);
            Alert.alert("Error", responseJson.message);
          }
        } catch (error) {
          console.log("Error: ", error);
          Analytics.trackEvent("Aadhaar|Otp|Error", {
            userId: id,
            error,
          });
          setVerifyMsg(error);
          setVerifyStatus("ERROR");
          setBackendPush(true);
          Alert.alert("Error", error);
        }
      })
      .catch((error) => {
        setVerifyMsg(error);
        setVerifyStatus("ERROR");
        setBackendPush(true);
        Analytics.trackEvent("Aadhaar|Otp|Error", {
          userId: id,
          error,
        });
        Alert.alert("Error", error);
      });
  };
  return (
    <PrimaryButton
      title={loading ? "Verifying" : props.title || "Continue"}
      disabled={loading}
      loading={loading}
      onPress={() => {
        goForFetch();
      }}
    />
  );
};

export default AadhaarOtpApi;
