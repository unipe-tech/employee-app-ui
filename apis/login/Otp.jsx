import { useNavigation } from "@react-navigation/core";
import Analytics from "appcenter-analytics";
import { useEffect, useState } from "react";
import { Alert, Text } from "react-native";
import { useDispatch, useSelector } from "react-redux";
import PrimaryButton from "../../components/atoms/PrimaryButton";
import { putBackendData } from "../../services/employees/employeeServices";
import { sendSmsVerification } from "../../services/otp/Gupshup/services";
import {
  addOnboarded,
  addToken,
  addUnipeEmployeeId
} from "../../store/slices/authSlice";
import { resetTimer } from "../../store/slices/timerSlice";
import { styles } from "../../styles";

const Otp = ({ disabled, type }) => {
  const dispatch = useDispatch();
  const navigation = useNavigation();

  const authSlice = useSelector((state) => state.auth);
  const [loading, setLoading] = useState(false);
  const [onboarded, setOnboarded] = useState(authSlice?.onboarded);
  const phoneNumber = authSlice?.phoneNumber;
  const [token, setToken] = useState(authSlice?.token);
  const [unipeEmployeeId, setUnipeEmployeeId] = useState(
    authSlice?.unipeEmployeeId
  );

  useEffect(() => {
    dispatch(addToken(token));
  }, [token]);

  useEffect(() => {
    dispatch(addOnboarded(onboarded));
  }, [onboarded]);

  useEffect(() => {
    dispatch(addUnipeEmployeeId(unipeEmployeeId));
  }, [unipeEmployeeId]);

  const SendOtp = (type) => {
    if (type === "login") {
      setLoading(true);
      dispatch(resetTimer());
      var fullPhoneNumber = `+91${phoneNumber}`;
      putBackendData({
        data: { number: fullPhoneNumber },
        xpath: "mobile",
        token: token,
      })
        .then((res) => {
          console.log("LoginScreen res.data: ", res.data);
          if (res.data.status === 200) {
            setOnboarded(res.data.body.onboarded);
            setToken(res.data.body.token);
            setUnipeEmployeeId(res.data.body.unipeEmployeeId);
            sendSmsVerification(phoneNumber)
              .then((result) => {
                console.log("sendSmsVerification result: ", result);
                if (result["response"]["status"] === "success") {
                  setLoading(false);
                  Analytics.trackEvent("LoginScreen|SendSms|Success", {
                    unipeEmployeeId: unipeEmployeeId,
                  });
                  navigation.navigate("Otp");
                } else {
                  setLoading(false);
                  Alert.alert(
                    result["response"]["status"],
                    result["response"]["details"]
                  );
                  Analytics.trackEvent("LoginScreen|SendSms|Error", {
                    unipeEmployeeId: unipeEmployeeId,
                    error: result["response"]["details"],
                  });
                }
              })
              .catch((error) => {
                console.log("sendSmsVerification result: ", error.toString());
                setLoading(false);
                Alert("Error", error.toString());
                Analytics.trackEvent("LoginScreen|SendSms|Error", {
                  unipeEmployeeId: unipeEmployeeId,
                  error: error.toString(),
                });
              });
            Analytics.trackEvent(`LoginScreen|SignIn|Success`, {
              unipeEmployeeId: res.data.body.id,
            });
          } else {
            setLoading(false);
            Alert.alert("Error", res.data["message"]);
            Analytics.trackEvent("LoginScreen|SignIn|Error", {
              phoneNumber: phoneNumber,
              error: res.data["message"],
            });
          }
        })
        .catch((error) => {
          console.log("LoginScreen res.data: ", error.toString());
          setLoading(false);
          Analytics.trackEvent("LoginScreen|SignIn|Error", {
            phoneNumber: phoneNumber,
            error: error.toString(),
          });
        });
    } else {
      sendSmsVerification(phoneNumber)
        .then((res) => {
          if (res["response"]["status"] === "success") {
            dispatch(resetTimer());
            Analytics.trackEvent("OTPScreen|SendSms|Success", {
              unipeEmployeeId: unipeEmployeeId,
            });
            Alert.alert("OTP resent successfully");
          } else {
            Analytics.trackEvent("OTPScreen|SendSms|Error", {
              unipeEmployeeId: unipeEmployeeId,
              error: res["response"]["details"],
            });
            Alert.alert(res["response"]["status"], res["response"]["details"]);
          }
        })
        .catch((error) => {
          console.log(error.toString());
          Analytics.trackEvent("OTPScreen|SendSms|Error", {
            unipeEmployeeId: unipeEmployeeId,
            error: error.toString(),
          });
          Alert.alert("Error", error.toString());
        });
    }
  };

  return (
    <>
      {type === "login" ? (
        <PrimaryButton
          title="Sign In"
          onPress={() => {
            SendOtp(type);
          }}
          disabled={disabled}
          loading={loading}
        />
      ) : (
        <Text
          style={styles.resendText}
          onPress={() => {
            SendOtp(type);
          }}
        >
          Resend
        </Text>
      )}
    </>
  );
};

export default Otp;
