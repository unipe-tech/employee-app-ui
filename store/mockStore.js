const store = {
  // Auth

  auth: {
    id: 1,
    phoneNumber: "9876543210",
    verifyStatus: "PENDING",
  },

  // Aadhaar

  aadhaar: {
    data: {
      aadhaar_data: {
        name: "Xyz garg",
      },
    },
    number: "789456123010",
    submitOTPtxnId: "987",
    verifyMsg: "This otp has been sent",
    verifyStatus: "PENDING",
  },

  // Bank

  bank: {
    accountNumber: "157894561230",
    ifsc: "INDB0000123",
    upi: "9876543210@upi",
    accountHolderName: "Xyz Goyal",
    verifyStatus: "PENDING",
    verifyMsg: "The verification is pending",
    bankName: "IndusInd Bank",
    bankBranch: "Mumbai",
    branchCity: "Mumbai",
  },

  // ESIC

  esic: {
    portal: { estCode: "", ipNumber: "" },
    familyDetails: {
      fatherHusband: { relation: "", name: "" },
      nominee: { relation: "", name: "" },
    },
    address: {
      present: { street: "", state: "", district: "", pincode: "" },
      permanent: { street: "", state: "", district: "", pincode: "" },
      nominee: { street: "", state: "", district: "", pincode: "" },
    },
  },

  // License

  license: {
    number: "",
    name: "",
    dob: "",
    bloodGroup: "",
    classes: {},
    rto: {},
    validity: {},
    verifyMsg: "",
    verifyStatus: "PENDING",
  },

  // Navigation

  navigation: { currentScreen: "Welcome" },

  // PAN

  pan: {
    dob: "02/02/2002",
    email: "xyz@unipe.co",
    gender: "Male",
    name: "Xyz Goyal",
    number: "ABCDE1234F",
    verifyMsg: "Verification Pending",
    verifyStatus: "PENDING",
  },

  // Profile

  profile: {
    maritalStatus: "Married",
    educationalQualification: "12th pass",
    alternatePhone: "+913216549870",
    email: "abc@unipe.co",
    selfie: "",
  },

  // Timer

  timer: { login: 2 * 60, aadhaar: 10 * 60 },
};

export default store;
