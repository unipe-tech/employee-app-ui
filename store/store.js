import { configureStore } from "@reduxjs/toolkit";
import authSlice from "./slices/authSlice";
import aadhaarSlice from "./slices/aadhaarSlice";
import panSlice from "./slices/panSlice";
import bankSlice from "./slices/bankSlice";
import esicSlice from "./slices/esicSlice";
import profileSlice from "./slices/profileSlice";
import navigationSlice from "./slices/navigationSlice";

export const store = configureStore({
  reducer: {
    auth: authSlice,
    aadhaar: aadhaarSlice,
    pan: panSlice,
    bank: bankSlice,
    profile: profileSlice,
    esic: esicSlice,
    navigation: navigationSlice
  },
});
