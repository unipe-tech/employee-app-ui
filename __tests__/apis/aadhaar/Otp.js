import React from "react";
import { NavigationContainer } from "@react-navigation/native";
import { Provider } from "react-redux";
import { store } from "../../../store/store";
import { shallow, mount } from "enzyme";
import Otp from "../../../apis/aadhaar/Otp";
import { render, fireEvent } from "@testing-library/react-native";
import { Button } from "@react-native-material/core";
import "isomorphic-fetch";
import { act } from "react-test-renderer";
import ApiView from "../../../apis/ApiView";

describe("AadhaarOTP API", () => {
  describe("UI rendering test", () => {
    it("testing UI rendering", () => {
      const tree = shallow(
        <Provider store={store}>
          <NavigationContainer>
            <Otp
              url="http://unipeapi.herokuapp.com/aadhaar-api/boson/generate-otp"
              data={{ aadhaar_number: "123456789101", consent: "Y" }}
              disabled={true}
            />
          </NavigationContainer>
        </Provider>
      );
      expect(tree).toMatchSnapshot();
    }, 8000);

    it("testing OTP component", () => {
      const goForFetch = jest.fn({
        fetchUrl:
          "http://unipeapi.herokuapp.com/aadhaar-api/boson/generate-otp",
        postData: { aadhaar_number: "123456789101", consent: "Y" },
      });
      const navigate = jest.fn();

      const component = render(
        <Provider store={store}>
          <NavigationContainer>
            <Otp
              url="http://unipeapi.herokuapp.com/aadhaar-api/boson/generate-otp/200/1001"
              data={{ aadhaar_number: "123456789101", consent: "Y" }}
              disabled={false}
              goForFetch={goForFetch}
              navigation={{ navigate }}
            />
          </NavigationContainer>
        </Provider>
      );
      const wrapper = shallow(
        <Provider store={store}>
          <NavigationContainer>
            <Otp
              url="http://unipeapi.herokuapp.com/aadhaar-api/boson/generate-otp/200/1001"
              data={{ aadhaar_number: "123456789101", consent: "Y" }}
              disabled={false}
              goForFetch={goForFetch}
              navigation={{ navigate }}
            />
          </NavigationContainer>
        </Provider>
      );
      fireEvent.press(component.getByTestId("apiButton"));
      act(async () => {
        expect(navigate).toHaveBeenCalledWith("AadhaarVerify");
      });
    });
  });
  describe("State tests", () => {
    let wrapper;
    let component;
    const setState = jest.fn();
    const useStateSpy = jest.spyOn(React, "useState");
    useStateSpy.mockImplementation((init) => [init, setState]);

    beforeEach(() => {
      const goForFetch = jest.fn();
      const navigate = jest.fn();
      wrapper = shallow(
        <Provider store={store}>
          <NavigationContainer>
            <Otp
              url="http://unipeapi.herokuapp.com/aadhaar-api/boson/generate-otp/200/1001"
              data={{ aadhaar_number: "123456789101", consent: "Y" }}
              disabled={false}
              goForFetch={goForFetch}
              navigation={{ navigate }}
            />
          </NavigationContainer>
        </Provider>
      );
      component = render(
        <Provider store={store}>
          <NavigationContainer>
            <Otp
              url="http://unipeapi.herokuapp.com/aadhaar-api/boson/generate-otp/200/1001"
              data={{ aadhaar_number: "123456789101", consent: "Y" }}
              disabled={false}
              goForFetch={goForFetch}
              navigation={{ navigate }}
            />
          </NavigationContainer>
        </Provider>
      );
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    describe("Should set Verify Message", () => {
      it("call setVerifyMessage and set verify message to sent", () => {
        expect(wrapper.find({ testID: "apiButton" })).toBeDefined();
        // wrapper.find({ testID: "apiButton" }).first().simulate("click");
        const goForFetch = jest.fn();
        fireEvent.press(component.getByText("Continue"));
        // expect(goForFetch).toBeCalled();
        // act(async () => {
        //   await expect(goForFetch).toHaveBeenCalledWith({
        //     fetchUrl:
        //       "http://unipeapi.herokuapp.com/aadhaar-api/boson/generate-otp/200/1001",
        //     postData: { aadhaar_number: "123456789101", consent: "Y" },
        //   });
        // });
        expect(setState).toHaveBeenCalled();
      });
    });
  });
});
