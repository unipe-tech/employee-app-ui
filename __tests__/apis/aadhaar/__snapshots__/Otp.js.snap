// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AadhaarOTP API UI rendering test testing UI rendering 1`] = `
<ContextProvider
  value={
    Object {
      "getServerState": undefined,
      "store": Object {
        "@@observable": [Function],
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Object {
        "addNestedSub": [Function],
        "getListeners": [Function],
        "handleChangeWrapper": [Function],
        "isSubscribed": [Function],
        "notifyNestedSubs": [Function],
        "trySubscribe": [Function],
        "tryUnsubscribe": [Function],
      },
    }
  }
>
  <ForwardRef(NavigationContainerInner)>
    <AadhaarOtpApi
      data={
        Object {
          "aadhaar_number": "123456789101",
          "consent": "Y",
        }
      }
      disabled={true}
      url="http://unipeapi.herokuapp.com/aadhaar-api/boson/generate-otp"
    />
  </ForwardRef(NavigationContainerInner)>
</ContextProvider>
`;
